resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -192,58 +225,94 @@ private[spark] class KubernetesClusterSchedulerBackend(
+            .addToLimits("memory", executorMemoryLimitQuantity)
+            .addToRequests("cpu", executorCpuQuantity)
+            .addToLimits("cpu", executorCpuQuantity)
+          .endResources()
+          .withEnv(requiredEnv.asJava)
+          .withPorts(requiredPorts.asJava)
+        .endContainer()
+      .endSpec()
+
+    var resolvedPodBuilder = shuffleServiceConfig.map { config =>
+      config.shuffleDirs.foldLeft(basePodBuilder) { (builder, dir) =>
+        builder.editSpec()
+          .addNewVolume()
+            .withName(FilenameUtils.getBaseName(dir))
+            .withNewHostPath().withPath(dir)
+            .endHostPath()
+          .endVolume()
+          .editFirstContainer()
+            .addNewVolumeMount()
+            .withName(FilenameUtils.getBaseName(dir))
+            .withMountPath(dir)
+            .endVolumeMount()
             .endContainer()
           .endSpec()
-        .done())
+      }
+    }.getOrElse(basePodBuilder)
+
+    try {
+      (executorId, kubernetesClient.pods().create(resolvedPodBuilder.build()))
     } catch {
       case throwable: Throwable =>
         logError("Failed to allocate executor pod.", throwable)
         throw throwable
     }
   }
 
-  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] = Future[Boolean] {
-    EXECUTOR_MODIFICATION_LOCK.synchronized {
-      if (requestedTotal > totalExpectedExecutors.get) {
-        logInfo(s"Requesting ${requestedTotal - totalExpectedExecutors.get}"
-          + s" additional executors, expecting total $requestedTotal and currently" +
-          s" expected ${totalExpectedExecutors.get}")
-        for (i <- 0 until (requestedTotal - totalExpectedExecutors.get)) {
-          runningExecutorPods += allocateNewExecutorPod()
+  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] =
+    Future[Boolean] {
+      EXECUTOR_MODIFICATION_LOCK.synchronized {
+        if (requestedTotal > totalExpectedExecutors.get) {
+          logInfo(s"Requesting ${requestedTotal - totalExpectedExecutors.get}"
+            + s" additional executors, expecting total $requestedTotal and currently" +
+            s" expected ${totalExpectedExecutors.get}")
+          for (i <- 0 until (requestedTotal - totalExpectedExecutors.get)) {
+            runningExecutorPods += allocateNewExecutorPod()
+          }
+        }
+        totalExpectedExecutors.set(requestedTotal)
+      }
+
+      var allPodsReady = true
+      for (executorPod <- runningExecutorPods.values) {
+        try {
+          val pod = kubernetesClient.pods().inNamespace(kubernetesNamespace)
+            .withName(executorPod.getMetadata().getName()).get()
+
+          allPodsReady &&= Readiness.isPodReady(pod)
+        } catch {
+          case throwable: Throwable =>
+            logWarning(s"Executor no longer exists.", throwable)
+            allPodsReady = false
         }
       }
-      totalExpectedExecutors.set(requestedTotal)
+
+      allPodsReady
```

Comment:
K8s API calls are asynchronous in and of themselves, as long as we're not specifically watching the resources for their readiness. I think it's actually semantically correct to try to request the executors here from the API server. If there's a network timeout or a failure then the state of the Future should reflect as such. `YarnSchedulerBackend` does this as well, in that it queries the application master with a message indicating the target number of executors that is desired. In client mode this could be a remote call.